File: javascript/binarySearch.js
/* 
DESCRIPTION
With binary search, you guess in the middle and eliminate half the remaining options every iteration.
Binary search only works when the array that is passed is sorted.

binary_search takes a sorted array and the item we are looking for as parameters.
In this case, we are looking for a number.

BIG O NOTATION
Logarithmic: O(log N)
*/

const binarySearch = (arr, num) => {
  // sort the list in case it is not already sorted
  arr = arr.sort((a, b) => a - b);
  let high = arr.length - 1;
  let low = 0;

  while (high >= low) {
    // take the ceiling value of mid in case the arr length is even.
    let mid = Math.ceil((high + low) / 2);
    let guess = arr[mid];

    // if the guess is correct, return the index
    if (guess === num) {
      return mid;
    }
    // if the guess is too high, adjust the ceiling
    if (guess > num) {
      high = mid - 1;
      // if the guess is too low
    } else {
      low = mid + 1;
    }
  }

  // if we exit the loop without returning the index, return null
  return null;
};

// TEST CASES
const arrayOne = [1, 3, 5, 7, 9];
const numOne = 3;
// => 1
const arrayTwo = [9, 921, 5, 102, 23, 99, 4, 1, 13];
const numTwo = 23;

console.log(binarySearch(arrayOne, numOne));

File: python/binary_search.py
# STRATEGY
# With binary search, you guess in the middle and eliminate half the remaining options every iteration.
# Binary search only works when the list that is passed is sorted.

# BIG O NOTATION
# Logarithmic: O(log N)

# DESCRIPTION
# binary_search takes a sorted list and the item we are looking for as parameters.
# In this case, we are looking for an integer.

def binary_search(list, item): 
    # sort the list in case it is not sorted
    list.sort()
    low = 0
    high = len(list) - 1
    while low <= high:
        mid = (low + high) / 2
        guess = list[mid]
        # found the item
        if guess == item:
            return mid
        # guess was too high
        if guess > item:
            high = mid - 1
        # guess was too low
        else:
            low =  mid + 1
    # if we exit the while loop, we return none
    return None

# TEST CASES
list_one = [1, 3, 5, 7, 9]
num_one = 3
# => 1
list_two = [9, 921, 5, 102, 23, 99, 4, 1, 13]
num_two = 23
# => 5

# OUTPUT
print(binary_search(list_two, num_two))